backend/bitboxbase/rpcclient/websocket.go

```diff
@@ -0,0 +1,238 @@
+// Copyright 2018 Shift Devices AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rpcclient
+
+import (
+	"crypto/rand"
+	"encoding/base32"
+	"fmt"
+
+	"github.com/digitalbitbox/bitbox-wallet-app/util/errp"
+	"github.com/digitalbitbox/bitbox-wallet-app/util/observable"
+	"github.com/digitalbitbox/bitbox-wallet-app/util/observable/action"
+	"github.com/flynn/noise"
+	"github.com/gorilla/websocket"
+	"github.com/sirupsen/logrus"
+)
+
+// runWebsocket sets up loops for sending/receiving, abstracting away the low level details about
+// pings, timeouts, clientection closing, etc.
+// It returns three channels: one to send messages to the client, one which notifies
+// when the clientection was closed and one to receive messages from the client
+//
+// Closing msg makes runWebsocket's goroutines quit.
+// The goroutines close client upon exit, due to a send/receive error or when msg is closed.
+// runWebsocket never closes msg.
+
+const (
+	opICanHasHandShaek          = "h"
+	opICanHasPairinVerificashun = "v"
+	opUCanHasDemo               = byte('d')
+	responseSuccess             = "\x00"
+	responseNeedsPairing        = "\x01"
+)
+
+func (rpcClient *RPCClient) runWebsocket(client *websocket.Conn) (send chan<- []byte, weHaveQuit chan<- struct{}, receive <-chan []byte, remoteHasQuit <-chan struct{}) {
```

Comment:
keep in mind that the named return values are actual variables, so better use them or don't write them ;)

it seems to all work, but the structure is a bit awkward, as it does too many things at once:
- reads/writes messages,
- parses and handles them,
- returns channels which are meant solely for RPC and nothing else. what messages are read/written should not be the concern of this function, but the callers'

it would be much nicer if this function only set up the websocket connection and exchanged messages, and left all parsing/handling to the function caller, with a signature like this:

`runWebsocket(onMessageReceive func([]byte) (send chan<- []byte, quit chan struct{}, error)`
where the read loop calls `onMessageReceive()` for the caller to check the op code and forward to the rpc client, and where the send channel read from to send messages, so the caller can send push notifications and rpc responses.

Then, msg parsing/handling is separated from communication, so it becomes trivial to replace the communication channel, unit test it, etc.